agent/consul/gateways/bind.go

```diff
@@ -0,0 +1,98 @@
+package gateways
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul/agent/configentry"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// referenceSet stores an O(1) accessible set of ResourceReference objects.
+type referenceSet = map[structs.ResourceReference]any
+
+// gatewayRefs maps a gateway kind/name to a set of resource references.
+type gatewayRefs = map[configentry.KindName][]structs.ResourceReference
+
+// BindRoutesToGateways takes a slice of bound API gateways and a slice of routes.
+// It iterates over the parent references for each route. These parents are gateways the
+// route should be bound to. If the parent matches a bound gateway, the route is bound to
+// the gateway. Otherwise, the route is unbound from the gateway if it was bound.
+//
+// The function returns a list of references to the modified BoundAPIGatewayConfigEntry objects,
+// a map of resource references to errors that occurred when they were attempted to be
+// bound to a gateway, and an error if the overall process was unsucessful.
+func BindRoutesToGateways(gateways []*structs.BoundAPIGatewayConfigEntry, routes []structs.BoundRoute) ([]*structs.BoundAPIGatewayConfigEntry, map[structs.ResourceReference]error, error) {
+	// errored stores the errors from events where a resource reference failed to bind to a gateway.
+	errored := make(map[structs.ResourceReference]error)
+
+	modified := make([]*structs.BoundAPIGatewayConfigEntry, 0, len(gateways))
+	for _, route := range routes {
+		routeModified, routeErrored, err := BindRouteToGateways(gateways, route)
+		if err != nil {
+			return nil, nil, err
+		}
+		modified = append(modified, routeModified...)
+		for ref, err := range routeErrored {
+			errored[ref] = err
+		}
+	}
+
+	return modified, errored, nil
+}
+
+// BindRouteToGateways takes a slice of bound API gateways and a route.
+// It iterates over the parent references for the given route. These parents are gateways the
+// route should be bound to. If the parent matches a bound gateway, the route is bound to the
+// gateway. Otherwise, the route is unbound from the gateway if it was bound.
```

Comment:
I think this helps clarify the meaning a bit since we would never bind and then immediately unbind in the same function call
```suggestion
// gateway. Otherwise, the route is unbound from the gateway if it was previously bound.
```